### ISSUE TITLE
Hanging after uploading of powershell provisioning script

### ISSUE THREAD

**Issue by datallah** [hashicorp/packer#4891]
I'm experiencing an issue where the packer build hangs trying to run a powershell provisioning script.
Based on the debug log, it's hanging after uploading the script, but before running it:

```
2017/05/15 15:41:48 ui: ESC[1;32m==> amazon-ebs: Machine successfully restarted, moving onESC[0m
2017/05/15 15:41:48 ui: ESC[1;32m==> amazon-ebs: Provisioning with Powershell...ESC[0m
2017/05/15 15:41:48 ui: ESC[1;32m==> amazon-ebs: Provisioning with shell script: win_ec2config_reset.ps1ESC[0m
2017/05/15 15:41:48 packer: 2017/05/15 15:41:48 Opening win_ec2config_reset.ps1 for reading
2017/05/15 15:41:48 packer: 2017/05/15 15:41:48 Building command line for: if (Test-Path variable:global:ProgressPreference){$ProgressPreference='SilentlyContinue'};$env:PACKER_BUILDER_TYPE="amazon-ebs"; $env:PACKER_BUILD_NAME="amazon-ebs"; &'c:/Windows/Temp/script.ps1';exit $LastExitCode
2017/05/15 15:41:48 packer: 2017/05/15 15:41:48 [INFO] 1083 bytes written for 'uploadData'
2017/05/15 15:41:48 [INFO] 1083 bytes written for 'uploadData'
2017/05/15 15:41:48 packer: 2017/05/15 15:41:48 Uploading file to 'c:/Windows/Temp/script.ps1'
2017/05/15 15:41:49 packer: #< CLIXML
```

It'll hang at this point until I cancel.

An interesting note is that it isn't the first script that is uploaded, it's the second time I'm trying to run a powershell provision after a reboot.

This is when creating a Windows 2008r2 AWS image.

This is similar to issue #3667.

The latest attempt was using Packer 1.0.0, but I have had the same issue with 0.12.2.
Packer is running on a amd64 CentOS Linux 6.8 host.

The example along with the debug log is at https://gist.github.com/datallah/b09092b137428de1f3774e593f69c90b
This is using ami-06d34c66 which is the us-west-2 2017.04.12 AWS Windows_Server-2008-R2_SP1-English-64Bit-Base image.

I ran it as follows:
```
export AWS_ACCESS_KEY=????
export AWS_SECRET_KEY=???
export PACKER_LOG=1
export PACKER_LOG_PATH=./packer_debug.log

packer build -var vpc_id=vpc-??????? -var subnet_id=subnet-?????? -var template_name=$(date +%Y-%m-%d_%H-%M-%S) -debug -only=amazon-ebs test.json
```

**Comment by taliesins** [comment]
@datallah these normally turn out being related to the fact that we continue provisioning before machine has completely finished being sysprepped. 

I had issues with earlier releases of a Windows 10 iso that i never resolved. Using a newer iso seemed to do the trick.

My sysprep scripts even go to the trouble of installing all Windows updates before enabling firewall and then enabling winrm. Check out [https://github.com/taliesins/packer-baseboxes/blob/master/windows/windows-10-amd64/Autounattend.xml.template#L232](https://github.com/taliesins/packer-baseboxes/blob/master/windows/windows-10-amd64/Autounattend.xml.template#L232)

Lookout for:

- Hanging at different point during the provisioning process (signs that we are connecting before job is done)
- Provisioning scripts executed in parallel
- Scripts restart machine during provisioning
- Scripts change winrm settings
- Scripts change firewall settings
- Scripts change network settings

**Comment by datallah** [comment]
@taliesins thanks for the notes.

In the example that I posted, there are 2 restarts, and it's after the 2nd restart that the hang happens.
There isn't anything done between the 2 restarts that seems at all questionable (it's the bootstrap.ps1 in the gist)

There are some changes to WinRM and the Firewall Settings, but, similar to how you're making the changes early on in sysprep, in my case, it's done in the instancedata script, and before any of the provisioning steps.

I tried adding a delay after the restart and before the provisioning step, but that doesn't seem to have any effect, it still hangs in exactly the same way.

Unfortunately I'm stuck with Windows 2008 and I'm already using the latest Amazon AMI, so it doesn't look like there's much I can do on that front.

**Comment by SwampDragons** [comment]
I've reproduced.  Weirdly, the build doesn't hang if you remove the line `Add-Content $profile "`nSet-PSDebug -Trace 1`n"` from your bootstrap.ps1.

I'm not sure why this line is causing trouble.  I'll try to investigate further.

**Comment by SwampDragons** [comment]
I've found a fix and submitted a PR here: https://github.com/packer-community/winrmcp/pull/27

**Comment by diegogoding** [comment]
@SwampDragons This fix appears to be breaking Windows 2016 builds under specific circumstances. I done some regression testing, prior to the PR: [https://github.com/packer-community/winrmcp/pull/27](url) it works fine. Switching to the PR version it then breaks. 
It appears to break on empty files, in this case an empty __init__.py
```
found error: Error removing temporary file $env:TEMP\winrmcp-ddcd4032-b8f8-4f2d-5bfa-eb8fd5e80a5c.tmp: cleanup operation returned code=1
```

Here are the logs.
```
2017/12/31 17:35:05 ___go_build_github_com_hashicorp_packer: 2017/12/31 17:35:05 Moving file from $env:TEMP\winrmcp-ddcd4032-b8f8-4f2d-5bfa-eb8fd5e80a5c.tmp to c:\some\python\module\__init__.py
2017/12/31 17:35:07 ___go_build_github_com_hashicorp_packer: #< CLIXML
2017/12/31 17:35:07 ___go_build_github_com_hashicorp_packer: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>2017/12/31 17:35:07 Removing temporary file $env:TEMP\winrmcp-ddcd4032-b8f8-4f2d-5bfa-eb8fd5e80a5c.tmp
2017/12/31 17:35:09 ___go_build_github_com_hashicorp_packer: #< CLIXML
2017/12/31 17:35:09 [INFO] (telemetry) ending file
2017/12/31 17:35:09 [INFO] (telemetry) found error: Error removing temporary file $env:TEMP\winrmcp-ddcd4032-b8f8-4f2d-5bfa-eb8fd5e80a5c.tmp: cleanup operation returned code=1
```
Starting getting this error on packer versions greater than 1.1.0. And I have isolate to this commit. If I put a little more text in the file, just a short comment snippet, no error.

**Comment by SwampDragons** [comment]
Would you mind opening a new ticket for me with this information so I can track it separately from the original issue?

**Comment by diegogoding** [comment]
Sure, the ticket is here: [issues: 5752](https://github.com/hashicorp/packer/issues/5752)
